packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java

```diff
@@ -24,14 +30,108 @@
+
+      InputConnection result = super.onCreateInputConnection(outAttrs);
+      if (result != null) {
+        // This should never happen.
+        Log.wtf(TAG, "Failed unexpectedly creating a webview input connection.");
+      }
+
+      final InputMethodManager imm =
+          (InputMethodManager) getContext().getSystemService(INPUT_METHOD_SERVICE);
+      proxyAdapterView.requestFocus();
+      final View containerView = this;
+
+      // This is the crucial trick that gets the InputConnection creation to happen on the correct
+      // thread. https://cs.chromium.org/chromium/src/content/public/android/java/src/org/chromium/content/browser/input/ThreadedInputConnectionFactory.java?l=169.
+      this.post(
+          new Runnable() {
+            @Override
+            public void run() {
+              // This is a hack to make InputMethodManager believe that the proxy view now has a focus.
+              // As a result, InputMethodManager will think that mProxyView is focused, and will call
+              // getHandler() of the view when creating input connection.
+
+              // Step 1: Set proxyAdapterView as InputMethodManager#mNextServedView. This does not
+              // affect the real window focus.
+              proxyAdapterView.onWindowFocusChanged(true);
+
+              // Step 2: Have InputMethodManager focus in on containerView. As a result, IMM will call
+              // onCreateInputConnection() on proxyAdapterView on the same thread as
+              // proxyAdapterView.getHandler(). It will also call subsequent InputConnection methods on
+              // this IME thread.
+              imm.isActive(containerView);
+            }
+          });
+      return null;
+    }
+
+    @Override
+    public boolean checkInputConnectionProxy(final View view) {
+      View previousProxy = threadedInputConnectionProxyView;
+      threadedInputConnectionProxyView = view;
+      if (previousProxy != view) {
+        proxyAdapterView =
+            new ThreadedInputConnectionProxyAdapterView(
+                /*containerView=*/ flutterView,
+                /*targetView=*/ view,
+                /*imeHandler=*/ view.getHandler());
+        final View container = this;
+        proxyAdapterView.requestFocus();
+        post(
+            new Runnable() {
+              @Override
+              public void run() {
+                InputMethodManager imm =
+                    (InputMethodManager) getContext().getSystemService(INPUT_METHOD_SERVICE);
+                imm.restartInput(container);
+              }
+            });
+      }
+      return super.checkInputConnectionProxy(view);
+    }
+  }
 
   @SuppressWarnings("unchecked")
-  FlutterWebView(Context context, BinaryMessenger messenger, int id, Map<String, Object> params) {
-    webView = new WebView(context);
+  FlutterWebView(
+      Context context,
+      BinaryMessenger messenger,
+      int id,
+      Map<String, Object> params,
+      final View flutterView) {
```

Comment:
With the new Android embedding the FlutterView may come and go, maybe better to prepare for that.
Actually I'm not sure how can a plugin get a reference to the current FlutterView with the new embedding. @matthew-carroll do you know if that's currently supported?

Reply:
The new embedding does not allow access to the FlutterView for plugins. As you mentioned, this is because FlutterEngines can move between FlutterViews, and even execute without a FlutterView. I haven't looked much at this PR, but odds are that either the things this code needs from FlutterView should be exposed as stable public APIs from the FlutterEngine, or we may want to restructure this code. 

At first blush, I'm noticing input connection management in this code, which could easily conflict with FlutterView's input connection management. We had the recent PR that seemed to suggest we may have the wrong responsibility boundaries across the embedding objects. Is this PR making an even stronger case for that? CC @xster

Reply:
This can definitely conflict with Flutter's input management, though I do not see any clean way to do it. If you have a better suggestion that will be great. I'm sharing a doc with more context on why these workarounds are needed.

Reply:
Even though this PR and related changes are workarounds for unusual conditions, it may be the case that we're accumulating enough of them that we should engineer the embedding to accommodate such things. I'll look forward to reading your doc to better understand the requirements and then I'll be happy to offer ideas for how we might clean up some architecture boundaries, even if the literal code remains unsightly.